Potential race condition in webflux transaction naming

During the totally unrelated PR #2832 one of the webflux tests failed due to an invalid transaction name:

```
expected: "PUT /functional/hello-mapping"
 but was: "unnamedPUT /functional/hello-mapping" 
```

The fact that this failure hasn't come up before makes it look like there is some kind of race-condition within the webflux instrumentation. Maybe between ending the transaction and updating the transaction name, because `end()` will set the name of the transaction to `unnamed` in case no name is present.

Full error log with stacktrace:
```
[2022-10-06T15:18:16.750Z] [ERROR] Tests run: 29, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 23.603 s <<< FAILURE! - in co.elastic.apm.agent.springwebflux.ServerFunctionalInstrumentationTest
[2022-10-06T15:18:16.750Z] [ERROR] methodMapping{String}[3]  Time elapsed: 0.506 s  <<< FAILURE!
[2022-10-06T15:18:16.750Z] org.opentest4j.AssertionFailedError: 
[2022-10-06T15:18:16.750Z] 
[2022-10-06T15:18:16.750Z] expected: "PUT /functional/hello-mapping"
[2022-10-06T15:18:16.750Z]  but was: "unnamedPUT /functional/hello-mapping"
[2022-10-06T15:18:16.750Z] 	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
[2022-10-06T15:18:16.750Z] 	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
[2022-10-06T15:18:16.750Z] 	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
[2022-10-06T15:18:16.750Z] 	at co.elastic.apm.agent.springwebflux.AbstractServerInstrumentationTest.checkTransaction(AbstractServerInstrumentationTest.java:455)
[2022-10-06T15:18:16.751Z] 	at co.elastic.apm.agent.springwebflux.AbstractServerInstrumentationTest.methodMapping(AbstractServerInstrumentationTest.java:229)
[2022-10-06T15:18:16.751Z] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[2022-10-06T15:18:16.751Z] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[2022-10-06T15:18:16.751Z] 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[2022-10-06T15:18:16.751Z] 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
[2022-10-06T15:18:16.751Z] 	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:727)
[2022-10-06T15:18:16.751Z] 	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
[2022-10-06T15:18:16.751Z] 	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
[2022-10-06T15:18:16.751Z] 	at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:156)
[2022-10-06T15:18:16.751Z] 	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:147)
[2022-10-06T15:18:16.751Z] 	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestTemplateMethod(TimeoutExtension.java:94)
[2022-10-06T15:18:16.751Z] 	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(InterceptingExecutableInvoker.java:103)
[2022-10-06T15:18:16.751Z] 	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.lambda$invoke$0(InterceptingExecutableInvoker.java:93)
[2022-10-06T15:18:16.751Z] 	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
[2022-10-06T15:18:16.751Z] 	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
[2022-10-06T15:18:16.751Z] 	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
[2022-10-06T15:18:16.751Z] 	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
[2022-10-06T15:18:16.751Z] 	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.invoke(InterceptingExecutableInvoker.java:92)
[2022-10-06T15:18:16.751Z] 	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.invoke(InterceptingExecutableInvoker.java:86)
[2022-10-06T15:18:16.751Z] 	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$7(TestMethodTestDescriptor.java:217)
[2022-10-06T15:18:16.751Z] 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
[2022-10-06T15:18:16.751Z] 	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:213)
[2022-10-06T15:18:16.751Z] 	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:138)
[2022-10-06T15:18:16.751Z] 	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:68)
[2022-10-06T15:18:16.751Z] 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:151)
[2022-10-06T15:18:16.751Z] 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
[2022-10-06T15:18:16.751Z] 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
[2022-10-06T15:18:16.751Z] 	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
[2022-10-06T15:18:16.751Z] 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
[2022-10-06T15:18:16.751Z] 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
[2022-10-06T15:18:16.752Z] 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
[2022-10-06T15:18:16.752Z] 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
[2022-10-06T15:18:16.752Z] 	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:35)
[2022-10-06T15:18:16.752Z] 	at org.junit.platform.engine.support.hierarchical.NodeTestTask$DefaultDynamicTestExecutor.execute(NodeTestTask.java:226)
[2022-10-06T15:18:16.752Z] 	at org.junit.platform.engine.support.hierarchical.NodeTestTask$DefaultDynamicTestExecutor.execute(NodeTestTask.java:204)
[2022-10-06T15:18:16.752Z] 	at org.junit.jupiter.engine.descriptor.TestTemplateTestDescriptor.execute(TestTemplateTestDescriptor.java:142)
[2022-10-06T15:18:16.752Z] 	at org.junit.jupiter.engine.descriptor.TestTemplateTestDescriptor.lambda$execute$2(TestTemplateTestDescriptor.java:110)
[2022-10-06T15:18:16.752Z] 	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
[2022-10-06T15:18:16.752Z] 	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
[2022-10-06T15:18:16.752Z] 	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
[2022-10-06T15:18:16.752Z] 	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
[2022-10-06T15:18:16.752Z] 	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
[2022-10-06T15:18:16.752Z] 	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
[2022-10-06T15:18:16.752Z] 	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
[2022-10-06T15:18:16.752Z] 	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
[2022-10-06T15:18:16.752Z] 	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1655)
[2022-10-06T15:18:16.752Z] 	at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:658)
[2022-10-06T15:18:16.752Z] 	at java.base/java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:274)
[2022-10-06T15:18:16.752Z] 	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
[2022-10-06T15:18:16.752Z] 	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
[2022-10-06T15:18:16.752Z] 	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
[2022-10-06T15:18:16.752Z] 	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1655)
[2022-10-06T15:18:16.752Z] 	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
[2022-10-06T15:18:16.752Z] 	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
[2022-10-06T15:18:16.752Z] 	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
[2022-10-06T15:18:16.752Z] 	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
[2022-10-06T15:18:16.752Z] 	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
[2022-10-06T15:18:16.752Z] 	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
[2022-10-06T15:18:16.752Z] 	at java.base/java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:274)
[2022-10-06T15:18:16.752Z] 	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1655)
[2022-10-06T15:18:16.752Z] 	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
[2022-10-06T15:18:16.752Z] 	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
[2022-10-06T15:18:16.752Z] 	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
[2022-10-06T15:18:16.753Z] 	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
[2022-10-06T15:18:16.753Z] 	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
[2022-10-06T15:18:16.753Z] 	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
[2022-10-06T15:18:16.753Z] 	at org.junit.jupiter.engine.descriptor.TestTemplateTestDescriptor.execute(TestTemplateTestDescriptor.java:110)
[2022-10-06T15:18:16.753Z] 	at org.junit.jupiter.engine.descriptor.TestTemplateTestDescriptor.execute(TestTemplateTestDescriptor.java:44)
[2022-10-06T15:18:16.753Z] 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:151)
[2022-10-06T15:18:16.753Z] 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
[2022-10-06T15:18:16.753Z] 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
[2022-10-06T15:18:16.753Z] 	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
[2022-10-06T15:18:16.753Z] 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
[2022-10-06T15:18:16.753Z] 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
[2022-10-06T15:18:16.753Z] 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
[2022-10-06T15:18:16.753Z] 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
[2022-10-06T15:18:16.753Z] 	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
[2022-10-06T15:18:16.753Z] 	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
[2022-10-06T15:18:16.753Z] 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
[2022-10-06T15:18:16.753Z] 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
[2022-10-06T15:18:16.753Z] 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
[2022-10-06T15:18:16.753Z] 	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
[2022-10-06T15:18:16.753Z] 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
[2022-10-06T15:18:16.753Z] 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
[2022-10-06T15:18:16.753Z] 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
[2022-10-06T15:18:16.753Z] 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
[2022-10-06T15:18:16.753Z] 	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
[2022-10-06T15:18:16.753Z] 	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
[2022-10-06T15:18:16.753Z] 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
[2022-10-06T15:18:16.753Z] 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
[2022-10-06T15:18:16.753Z] 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
[2022-10-06T15:18:16.753Z] 	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
[2022-10-06T15:18:16.753Z] 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
[2022-10-06T15:18:16.753Z] 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
[2022-10-06T15:18:16.754Z] 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
[2022-10-06T15:18:16.754Z] 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
[2022-10-06T15:18:16.754Z] 	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:35)
[2022-10-06T15:18:16.754Z] 	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
[2022-10-06T15:18:16.754Z] 	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:54)
[2022-10-06T15:18:16.754Z] 	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:220)
[2022-10-06T15:18:16.754Z] 	at org.junit.platform.launcher.core.DefaultLauncher.lambda$execute$6(DefaultLauncher.java:188)
[2022-10-06T15:18:16.754Z] 	at org.junit.platform.launcher.core.DefaultLauncher.withInterceptedStreams(DefaultLauncher.java:202)
[2022-10-06T15:18:16.754Z] 	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:181)
[2022-10-06T15:18:16.754Z] 	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:128)
[2022-10-06T15:18:16.754Z] 	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invokeAllTests(JUnitPlatformProvider.java:150)
[2022-10-06T15:18:16.754Z] 	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invoke(JUnitPlatformProvider.java:124)
[2022-10-06T15:18:16.754Z] 	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:384)
[2022-10-06T15:18:16.754Z] 	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:345)
[2022-10-06T15:18:16.754Z] 	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:126)
[2022-10-06T15:18:16.754Z] 	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:418)
```

--- a/apm-agent-core/src/main/java/co/elastic/apm/agent/impl/error/ErrorCapture.java
+++ b/apm-agent-core/src/main/java/co/elastic/apm/agent/impl/error/ErrorCapture.java
@@ -19,7 +19,6 @@
 package co.elastic.apm.agent.impl.error;
 
 import co.elastic.apm.agent.configuration.CoreConfiguration;
-import co.elastic.apm.agent.impl.ActivationListener;
 import co.elastic.apm.agent.impl.ElasticApmTracer;
 import co.elastic.apm.agent.impl.context.TransactionContext;
 import co.elastic.apm.agent.impl.stacktrace.StacktraceConfiguration;
@@ -34,7 +33,6 @@ import co.elastic.apm.agent.sdk.logging.LoggerFactory;
 
 import javax.annotation.Nullable;
 import java.util.Collection;
-import java.util.List;
 
 
 /**
@@ -265,7 +263,7 @@ public class ErrorCapture implements Recyclable {
         transactionInfo.isSampled = transactionSampled;
     }
 
-    public void setTransactionName(@Nullable StringBuilder name) {
+    public void setTransactionName(@Nullable CharSequence name) {
         transactionInfo.name.setLength(0);
         transactionInfo.name.append(name);
     }

--- a/apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/AbstractSpan.java
+++ b/apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/AbstractSpan.java
@@ -245,8 +245,12 @@ public abstract class AbstractSpan<T extends AbstractSpan<T>> implements Recycla
     /**
      * Only intended to be used by {@link co.elastic.apm.agent.report.serialize.DslJsonSerializer}
      */
-    public StringBuilder getNameForSerialization() {
-        return name;
+    public CharSequence getNameForSerialization() {
+        if (name.length() == 0) {
+            return "unnamed";
+        } else {
+            return name;
+        }
     }
 
     /**
@@ -324,8 +328,12 @@ public abstract class AbstractSpan<T extends AbstractSpan<T>> implements Recycla
     }
 
     public T appendToName(CharSequence cs, int priority) {
+        return appendToName(cs, priority, 0, cs.length());
+    }
+
+    public T appendToName(CharSequence cs, int priority, int startIndex, int endIndex) {
         if (priority >= namePriority) {
-            this.name.append(cs);
+            this.name.append(cs, startIndex, endIndex);
             this.namePriority = priority;
         }
         return thiz();
@@ -583,9 +591,6 @@ public abstract class AbstractSpan<T extends AbstractSpan<T>> implements Recycla
     public final void end(long epochMicros) {
         if (!finished) {
             this.endTimestamp.set(epochMicros);
-            if (name.length() == 0) {
-                name.append("unnamed");
-            }
             childDurations.onSpanEnd(epochMicros);
 
             type = normalizeType(type);

--- a/apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/StackFrame.java
+++ b/apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/StackFrame.java
@@ -44,10 +44,11 @@ public class StackFrame {
         return methodName;
     }
 
-    public void appendSimpleClassName(StringBuilder sb) {
+    public int getSimpleClassNameOffset() {
         if (className != null) {
-            sb.append(className, className.lastIndexOf('.') + 1, className.length());
+            return className.lastIndexOf('.') + 1;
         }
+        return 0;
     }
 
     public void appendFileName(StringBuilder replaceBuilder) {

--- a/apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/Transaction.java
+++ b/apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/Transaction.java
@@ -410,13 +410,10 @@ public class Transaction extends AbstractSpan<Transaction> {
     public CharSequence getNameForSerialization() {
         WildcardMatcher match = WildcardMatcher.anyMatch(coreConfig.getTransactionNameGroups(), this.name);
         if (match != null) {
-            CharSequence matchName = match.toString();
             this.name.setLength(0);
-            this.name.append(matchName);
-            return matchName;
-        } else {
-            return super.getNameForSerialization();
+            this.name.append(match);
         }
+        return super.getNameForSerialization();
     }
 
     @Override

--- a/apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/Transaction.java
+++ b/apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/Transaction.java
@@ -407,14 +407,16 @@ public class Transaction extends AbstractSpan<Transaction> {
     }
 
     @Override
-    public StringBuilder getNameForSerialization() {
-        StringBuilder name = this.name;
-        WildcardMatcher match = WildcardMatcher.anyMatch(coreConfig.getTransactionNameGroups(), name);
+    public CharSequence getNameForSerialization() {
+        WildcardMatcher match = WildcardMatcher.anyMatch(coreConfig.getTransactionNameGroups(), this.name);
         if (match != null) {
-            name.setLength(0);
-            name.append(match);
+            CharSequence matchName = match.toString();
+            this.name.setLength(0);
+            this.name.append(matchName);
+            return matchName;
+        } else {
+            return super.getNameForSerialization();
         }
-        return name;
     }
 
     @Override

--- a/apm-agent-core/src/test/java/co/elastic/apm/agent/impl/ElasticApmTracerTest.java
+++ b/apm-agent-core/src/test/java/co/elastic/apm/agent/impl/ElasticApmTracerTest.java
@@ -307,7 +307,7 @@ class ElasticApmTracerTest {
             .isTrue();
         assertThat(error.getTransactionInfo().isSampled()).isEqualTo(sampled);
         if (!transaction.getNameAsString().isEmpty()) {
-            assertThat(error.getTransactionInfo().getName()).isEqualTo(transaction.getNameAsString());
+            assertThat(error.getTransactionInfo().getName().toString()).isEqualTo(transaction.getNameAsString());
         }
         assertThat(error.getTransactionInfo().getType()).isEqualTo(transaction.getType());
         return error;

--- a/apm-agent-core/src/test/java/co/elastic/apm/agent/impl/transaction/SpanTest.java
+++ b/apm-agent-core/src/test/java/co/elastic/apm/agent/impl/transaction/SpanTest.java
@@ -61,7 +61,7 @@ class SpanTest {
             .withUser("readonly_user");
         span.resetState();
         assertThat(span.getContext().hasContent()).isFalse();
-        assertThat(span.getNameAsString()).isNullOrEmpty();
+        assertThat(span.getNameAsString()).isEqualTo("unnamed");
         assertThat(span.getType()).isNull();
         assertThat(span.getSubtype()).isNull();
         assertThat(span.getAction()).isNull();

--- a/apm-agent-plugins/apm-profiling-plugin/src/main/java/co/elastic/apm/agent/profiler/CallTree.java
+++ b/apm-agent-plugins/apm-profiling-plugin/src/main/java/co/elastic/apm/agent/profiler/CallTree.java
@@ -20,6 +20,7 @@ package co.elastic.apm.agent.profiler;
 
 import co.elastic.apm.agent.collections.LongList;
 import co.elastic.apm.agent.impl.ElasticApmTracer;
+import co.elastic.apm.agent.impl.transaction.AbstractSpan;
 import co.elastic.apm.agent.impl.transaction.Span;
 import co.elastic.apm.agent.impl.transaction.StackFrame;
 import co.elastic.apm.agent.impl.transaction.TraceContext;
@@ -403,7 +404,12 @@ public class CallTree implements Recyclable {
             .withSubtype("inferred")
             .withChildIds(childIds);
 
-        frame.appendSimpleClassName(span.getNameForSerialization());
+        String classFqn = frame.getClassName();
+        if (classFqn != null) {
+            span.appendToName(classFqn, AbstractSpan.PRIO_DEFAULT, frame.getSimpleClassNameOffset(), classFqn.length());
+        } else {
+            span.appendToName("null");
+        }
         span.appendToName("#");
         span.appendToName(frame.getMethodName());
 

